daemon/containerd/soft_delete.go

```diff
@@ -0,0 +1,73 @@
+package containerd
+
+import (
+	"context"
+
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
+	"github.com/docker/docker/errdefs"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+)
+
+// softImageDelete deletes the image, making sure that there are other images
+// that reference the content of the deleted image.
+// If no other image exists, a dangling one is created.
+func (i *ImageService) softImageDelete(ctx context.Context, img containerdimages.Image) error {
+	is := i.client.ImageService()
+
+	logger := logrus.WithFields(logrus.Fields{
+		"name":   img.Name,
+		"digest": img.Target.Digest.String(),
+	})
+
+	// If the image already exists, persist it as dangling image
+	// but only if no other image has the same target.
+	digest := img.Target.Digest.String()
+	imgs, err := is.List(ctx, "target.digest=="+digest)
+	if err != nil {
+		logger.WithField("digest", digest).WithError(err).Debug("failed to check if there are images targeting this digest")
+		return errdefs.System(err)
+	}
+
+	// From this point explicitly ignore the passed context
+	// and don't allow to interrupt this in the middle.
+	ctx = context.Background()
+
+	// Create dangling image if this is the last image pointing to this target.
+	if len(imgs) == 1 {
+		danglingImage := img
+
+		danglingImage.Name = danglingImageName(img.Target.Digest)
+		delete(danglingImage.Labels, "io.containerd.image.name")
+		delete(danglingImage.Labels, "org.opencontainers.image.ref.name")
+
+		_, err = is.Create(ctx, danglingImage)
+
+		// Error out in case we couldn't persist the old image.
+		// If it already exists, then just continue.
+		if err != nil && !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Error("failed to create a dangling image for the image being replaced")
+			return errdefs.System(err)
+		}
+	}
+
+	// Free the target name.
+	err = is.Delete(ctx, img.Name)
+	if err != nil {
+		if !cerrdefs.IsNotFound(err) {
+			logger.WithError(err).Error("failed to delete image which existed a moment before")
+			return errdefs.System(err)
+		}
+	}
+
+	return nil
+}
+
+func isDanglingImage(image containerdimages.Image) bool {
+	return image.Name == danglingImageName(image.Target.Digest)
+}
+
+func danglingImageName(digest digest.Digest) string {
+	return "dangling@" + digest.String()
```

Comment:
Should we make this something like `moby-dangling@xxx` so it's obvious even in `ctr` what this is? :innocent:

Reply:
Sounds good!